File: Tests/Tests/STPBankAccountFunctionalTest.m
//
//  STPBankAccountFunctionalTest.m
//  Stripe
//
//  Created by Charles Scalesse on 10/2/14.
//
//

@import XCTest;
@import StripeCoreTestUtils;


#import "STPTestingAPIClient.h"


@interface STPBankAccountFunctionalTest : XCTestCase
@end

@implementation STPBankAccountFunctionalTest

- (void)testCreateAndRetreiveBankAccountToken {
    STPBankAccountParams *bankAccount = [[STPBankAccountParams alloc] init];
    bankAccount.accountNumber = @"000123456789";
    bankAccount.routingNumber = @"110000000";
    bankAccount.country = @"US";
    bankAccount.accountHolderName = @"Jimmy bob";
    bankAccount.accountHolderType = STPBankAccountHolderTypeCompany;

    STPAPIClient *client = [[STPAPIClient alloc] initWithPublishableKey:STPTestingDefaultPublishableKey];

    XCTestExpectation *expectation = [self expectationWithDescription:@"Bank account creation"];
    [client createTokenWithBankAccount:bankAccount
                            completion:^(STPToken *token, NSError *error) {
                                [expectation fulfill];
                                XCTAssertNil(error, @"error should be nil %@", error.localizedDescription);
                                XCTAssertNotNil(token, @"token should not be nil");

                                XCTAssertNotNil(token.tokenId);
                                XCTAssertEqual(token.type, STPTokenTypeBankAccount);
                                XCTAssertNotNil(token.bankAccount.stripeID);
                                XCTAssertEqualObjects(@"STRIPE TEST BANK", token.bankAccount.bankName);
                                XCTAssertEqualObjects(@"6789", token.bankAccount.last4);
                                XCTAssertEqualObjects(@"Jimmy bob", token.bankAccount.accountHolderName);
                                XCTAssertEqual(token.bankAccount.accountHolderType, STPBankAccountHolderTypeCompany);
                            }];

    [self waitForExpectationsWithTimeout:TestConstants.STPTestingNetworkRequestTimeout handler:nil];
}

- (void)testInvalidKey {
    STPBankAccountParams *bankAccount = [[STPBankAccountParams alloc] init];
    bankAccount.accountNumber = @"000123456789";
    bankAccount.routingNumber = @"110000000";
    bankAccount.country = @"US";

    STPAPIClient *client = [[STPAPIClient alloc] initWithPublishableKey:@"not_a_valid_key_asdf"];

    XCTestExpectation *expectation = [self expectationWithDescription:@"Bad bank account creation"];

    [client createTokenWithBankAccount:bankAccount
                            completion:^(STPToken *token, NSError *error) {
                                [expectation fulfill];
                                XCTAssertNil(token, @"token should be nil");
                                XCTAssertNotNil(error, @"error should not be nil");
                            }];
    [self waitForExpectationsWithTimeout:TestConstants.STPTestingNetworkRequestTimeout handler:nil];
}

@end

File: Stripe/StripeiOSTests/STPBankAccountFunctionalTest.swift
//  Converted to Swift 5.8.1 by Swiftify v5.8.28463 - https://swiftify.com/
//
//  STPBankAccountFunctionalTest.m
//  Stripe
//
//  Created by Charles Scalesse on 10/2/14.
//
//

import StripeCoreTestUtils
import StripePaymentsTestUtils
import XCTest

class STPBankAccountFunctionalTest: STPNetworkStubbingTestCase {
    func testCreateAndRetreiveBankAccountToken() {
        let bankAccount = STPBankAccountParams()
        bankAccount.accountNumber = "000123456789"
        bankAccount.routingNumber = "110000000"
        bankAccount.country = "US"
        bankAccount.accountHolderName = "Jimmy bob"
        bankAccount.accountHolderType = STPBankAccountHolderType.company

        let client = STPAPIClient(publishableKey: STPTestingDefaultPublishableKey)

        let expectation = self.expectation(description: "Bank account creation")
        client.createToken(
            withBankAccount: bankAccount) { token, error in
            expectation.fulfill()
            XCTAssertNil(error, "error should be nil")
            XCTAssertNotNil(token, "token should not be nil")

            XCTAssertNotNil(token?.tokenId)
                XCTAssertEqual(token?.type, .bankAccount)
            XCTAssertNotNil(token?.bankAccount?.stripeID)
                XCTAssertEqual("STRIPE TEST BANK", token?.bankAccount?.bankName)
                XCTAssertEqual("6789", token?.bankAccount?.last4)
            XCTAssertEqual("Jimmy bob", token?.bankAccount?.accountHolderName)
            XCTAssertEqual(token?.bankAccount?.accountHolderType, STPBankAccountHolderType.company)
        }

        waitForExpectations(timeout: STPTestingNetworkRequestTimeout, handler: nil)
    }

    func testInvalidKey() {
        let bankAccount = STPBankAccountParams()
        bankAccount.accountNumber = "000123456789"
        bankAccount.routingNumber = "110000000"
        bankAccount.country = "US"

        let client = STPAPIClient(publishableKey: "not_a_valid_key_asdf")

        let expectation = self.expectation(description: "Bad bank account creation")

        client.createToken(
            withBankAccount: bankAccount) { token, error in
            expectation.fulfill()
            XCTAssertNil(token, "token should be nil")
            XCTAssertNotNil(error, "error should not be nil")
        }
        waitForExpectations(timeout: STPTestingNetworkRequestTimeout, handler: nil)
    }
}
